SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialFunctionalTest.cpp

```diff
@@ -196,8 +196,18 @@ void ASpatialFunctionalTest::LogStep(ELogVerbosity::Type Verbosity, const FStrin
 
 void ASpatialFunctionalTest::PrepareTest()
 {
-	StepDefinitions.Empty();
+	if (bFinishedTest && !bNotifyObserversCalled) // This happens when the world is not reset.
```

Comment:
Basically here tests can be started before we have called NotifyObservers for the previous test. This shouldn't happen often but if the worker that Unreal adds the delegate to changes then weird behavior can happen where we call NotifyObserver for another test due to an in-flight RPC

Reply:
+ the reason it was done like this is because NotifyObservers is the call which triggers the end of the test (and start of a new one, which is potentially the same ASpatialFunctionTest if the "runs" option is used)

Reply:
Although it seems worrying to me that a new test can be starting while the old one hasn't finished signalling that it has ended.
I was under the illusion that the end of the test would only occur if all workers have confirmed that it can end, but this is apparently not the case :(
Do we want to look into this more in the future?

Reply:
Yeah I think the crux of the problem is we distribute calling NotifyObservers and native Unreal (I assume) expects no further test code to run once the main test runner has called this function. So we could trigger this on one worker and then tell all the others to go and call it but it may have already actually ended (and restarted) by the time the message arrives. I think this is fine in single-server because the clients calling this doesn't trigger test end but I think I was seeing the responsible worker change in multi-server (I don't know how Unreal chooses which world to add the delegate to - I assume it's networked if the tests are running in other processes!?)

Perhaps we could figure out which worker is responsible for actually finishing the test and having that worker acquire consensus on test end before calling this it might work better. I did consider modifying the framework but this seems more risky.

It could be worth looking into further if this is just adding a tech debt workaround